### ISSUE TITLE
使用 nn.Graph 训练 vit base 速度慢于 pytorch

### ISSUE THREAD

**Issue by SunDoge** [Oneflow-Inc/oneflow#7290]
## Summary

在3090上对 pytorch 和 oneflow 的 vit_base_patch16_224 模型进行 benchmark, 主要测试训练速度，输入 fake image 和 fake label，使用 sgd 并 backward，最后都调用 `loss.item()` 进行同步，速度如下

| Framework | Speed |
| ------------ | ------- |
| pytorch | 200 img/s |
| oneflow | 170 img/s |
| oneflow w/ nn.LayerNorm | 180 img/s |

## Code to reproduce bug

[https://github.com/SunDoge/vit-bench](https://github.com/SunDoge/vit-bench)

## System Information

- What is your OneFlow installation (pip, source, dockerhub): pip
- OS: Ubuntu 20.04
- OneFlow version (run `python3 -m oneflow --doctor`): 0.6.0+cu112, 0.7.0.dev20220117+cu112
- Python version: 3.9.7
- CUDA driver version: Driver Version: 460.73.01    CUDA Version: 11.2
- GPU models: vit_base_patch16_224
- Other info:

**Comment by chengtbf** [comment]
您好，非常感谢您试用 oneflow~

我在 21 机器上做的本地测试。使用 https://github.com/SunDoge/vit-bench 的仓库代码。

测试环境是：
- OS: Ubuntu 20.04.1 LTS (GNU/Linux 5.8.0-59-generic x86_64)
- OneFlow version: version: 0.7.0+cu112.git.3e8be2e1e3
- Python version:  3.8.5 
- CUDA driver version: 460.73.01, CUDA Version: 11.2
- PyTorch Version: 1.10.1
- GPU: GeForce RTX 2080 (8G)

由于显存是 8G，所以我改 batch size 从 64 -> 32， oneflow 版本是 nightly， pytorch 是 1.10.1。如果直接跑您的代码：

```sh
chengcheng@oneflow-21:~ $ git clone git@github.com:SunDoge/vit-bench.git
chengcheng@oneflow-21:~ $ cd vit-bench/
chengcheng@oneflow-21:~/vit-bench $ vim bench_oneflow_vit.py 
chengcheng@oneflow-21:~/vit-bench $ python3 bench_oneflow_vit.py 
100%|██████████████████████████████████████████████████████████████████████████| 320/320 [00:15<00:00, 20.64it/s]
100%|████████████████████████████████████████████████████████████████████████| 3200/3200 [00:45<00:00, 70.15it/s]

chengcheng@oneflow-21:~/vit-bench $ python3 bench_pytorch_vit.py 
100%|██████████████████████████████████████████████████████████████████████████| 320/320 [00:04<00:00, 66.29it/s]
100%|████████████████████████████████████████████████████████████████████████| 3200/3200 [00:47<00:00, 67.82it/s]
```

oneflow 是比 torch 稍快 5% 的。 

nn.Graph 的 input 会稍稍影响流水，因为涉及到数据从 Python 传递到 Graph 内部，所以 Graph 的更高效做法是把 input 放在 Graph 内部，具体修改是：

```python
class VitTrainGraph(nn.Graph):

    def __init__(self, model, optimizer):
        super().__init__()
        self.model = model
        self.criterion = nn.CrossEntropyLoss()

        self.config.allow_fuse_add_to_output()
        self.config.allow_fuse_model_update_ops()
        self.add_optimizer(optimizer)

    def build(self):
        x = oneflow.rand(32, 3, 224, 224).to(oneflow.device('cuda'))
        y = oneflow.randint(0, 1000, (32,)).to(oneflow.device('cuda'))
        y_pred = self.model(x)
        loss = self.criterion(y_pred, y)
        loss.backward()
        return loss, y_pred
```

然后外面的代码去掉 x、y 的初始化即可。

这样会有一点提速：

```sh
chengcheng@oneflow-21:~/vit-bench $ python3 bench_oneflow_vit.py 
100%|██████████████████████████████████████████████████████████████████████████| 320/320 [00:11<00:00, 27.95it/s]
100%|████████████████████████████████████████████████████████████████████████| 3200/3200 [00:45<00:00, 71.03it/s]
```

在 2080 上，oneflow 和 torch 都可以打满 gpu（利用率均是 100%），性能差异的来源主要是具体 kernel 优化的快慢。


我们也有 3090 的机器。我稍后去 14 机器上再测试一次。 
跟用户的区别就是我本地用的 oneflow 的版本稍新一些。
（可以用这个方式安装：`python3 -m pip install --pre oneflow -f https://staging.oneflow.info/branch/master/cu112`）

**Comment by yuanms2** [comment]
这里使用的oneflow graph 也就是静态图模式，pytorch 使用torch script 对比才公平吧？

**Comment by chengtbf** [comment]
> 这里使用的oneflow graph 也就是静态图模式，pytorch 使用torch script 对比才公平吧？

确实，如果是比较 Graph 版本，跟 torch script 才是完全对齐的。 我们也可以先跑一个 OneFlow Eager 版本的结果，对比动态图的性能。

**Comment by SunDoge** [comment]
Hi, @yuanms2 @chengtbf  目前的 torch script 并没有做 optimization，所以并无加速，而且一般训练也不会使用 torch script。

@chengtbf 我在 TITAN XP 上得到了与 2080 类似的结果，所以可能是您说的 python 部分的 overhead。即
> nn.Graph 的 input 会稍稍影响流水，因为涉及到数据从 Python 传递到 Graph 内部

当 GPU 性能提升之后这个 overhead 可能会比较明显。

**Comment by chengtbf** [comment]
Graph 外部 input 对流水的影响我们后续会继续解决这个问题，尽量做到和 Graph 内部的 input 一样快。

如果 vit base 使用 OneFlow Eager 对比的话，性能差距更小一些，只领先 2% 左右

```
chengcheng@oneflow-21:~/vit-bench $ python3 bench_oneflow_vit.py 
100%|██████████████████████████████████████████████████████████████████████████| 320/320 [00:06<00:00, 51.79it/s]
100%|████████████████████████████████████████████████████████████████████████| 3200/3200 [00:46<00:00, 68.83it/s]
```

vit 这个模型我们后续再针对性的调优一下，应该还有提升的空间。

**Comment by yuanms2** [comment]
可以在3090 上页测试一下

**Comment by SunDoge** [comment]
> nn.Graph 的 input 会稍稍影响流水，因为涉及到数据从 Python 传递到 Graph 内部，所以 Graph 的更高效做法是把 input 放在 Graph 内部，具体修改是：

参考了这个修改之后，3090 上的速度为 395.83 img/s， 加速明显。但是我看了一下文档，按我的理解，只有基于 `flow.nn.OfrecordReader` 的数据读取方式可以作为静态图节点，基于 `flow.utils.data.DataLoader` 有办法优化吗？

**Comment by chenyaofo** [comment]
对Oneflow相对于PyTorch的速度提升比较感兴趣，这里提供一些V100和A100上的结果供大家讨论。根据@chengtbf的建议，已经将inputs放到了Graph内部，代码放在：https://gitee.com/chenyaofo/vit-bench

测试环境1 (V100)：

- OS: Ubuntu 20.04.3 LTS (GNU/Linux 4.15.0-45-generic x86_64)
- OneFlow version: version: 0.7.0.dev20220121+cu112
- Python version: 3.8.10
- CUDA driver version: 460.32.03, CUDA Version: 11.4
- PyTorch Version: 1.10.1+cu113
- GPU: Tesla V100-SXM2-32GB

测试结果：PyTorch 92 fps *vs.* OneFlow 103 fps，OneFlow 领先 12.0%
```
~/vit-bench# CUDA_VISIBLE_DEVICES=0 python bench_pytorch_vit.py
100%|█████████████████████████████████████████████████████████████████████████████████| 640/640 [00:08<00:00, 74.56it/s]
100%|███████████████████████████████████████████████████████████████████████████████| 6400/6400 [01:09<00:00, 92.31it/s]
~/vit-bench# CUDA_VISIBLE_DEVICES=0 python bench_oneflow_vit.py
libibverbs not available, ibv_fork_init skipped
100%|█████████████████████████████████████████████████████████████████████████████████| 640/640 [00:12<00:00, 52.89it/s]
100%|██████████████████████████████████████████████████████████████████████████████| 6400/6400 [01:02<00:00, 103.04it/s]
```


---

测试环境2 (A100)：

- OS: Ubuntu 18.04.5 LTS (GNU/Linux 4.15.0-156-generic x86_64)
- OneFlow version: version: 0.7.0.dev20220121+cu112
- Python version: 3.8.10
- CUDA driver version: 470.57.02, CUDA Version: 11.4
- PyTorch Version: 1.10.1+cu113
- GPU: NVIDIA A100-SXM4-40GB

测试结果：PyTorch 450 fps *vs.* OneFlow 473 fps，OneFlow 领先 5.1%
```
~/vit-bench$ CUDA_VISIBLE_DEVICES=0 python bench_pytorch_vit.py
100%|████████████████████████████████████████████████████████████████████████████████| 640/640 [00:01<00:00, 389.04it/s]
100%|██████████████████████████████████████████████████████████████████████████████| 6400/6400 [00:14<00:00, 449.52it/s]
~/vit-bench$ CUDA_VISIBLE_DEVICES=0 python bench_oneflow_vit.py
100%|█████████████████████████████████████████████████████████████████████████████████| 640/640 [00:06<00:00, 93.51it/s]
100%|██████████████████████████████████████████████████████████████████████████████| 6400/6400 [00:13<00:00, 473.07it/s]
```

---

个人杂谈：OneFlow在这个模型（VIT）上比PyTorch快5%-10%左右，这个领先幅度似乎还不够决定性（也许是我现在的姿势不对？我感觉这种静态图的模式应该快很多啊）。OneFlow对于我来说的两大亮点，第一是性能，第二是分布式设计（不过这一点暂时没有那么重要，对于小作坊PyTorhc的DistributedParallel已经够用了），希望OneFlow这两个亮点能越做越强吧。

**Comment by yuanms2** [comment]
如果pytorch和oneflow的kernel 实现都一模一样，区别就是底层的执行引擎，现在vit这样的模型上cuda kernel的排列已经很紧密了，反映在GPU利用率上就是差不多100%, 性能差异大概率来自cuda kernel的实现不一样，单卡情况在vit这个模型上提升空间不大，更大的空间只能是用张量编译器做更激进的kernel fusion。

另外还发现之前oneflow的vit 模型实现和pytorch 不一样，在pytorch里，计算q,k,v 时只需要1个矩阵乘，在oneflow的vit实现里是做3次矩阵乘，这样会慢一些，有一位同事正在修正vit的算法实现。

**Comment by chengtbf** [comment]
另外再补充一点，FP32 上调度开销的差距并不明显（因为计算相对任务重，Kernel 执行的时间相对长，调度开销的占比低），而 AMP 上的优化可能差异会明显一些（之前在 ResNet50 上的差距也可以供类比， FP32 只能领先 10%，但是 AMP 最多可以领先 80%）。vit、Swin-Transformer 是比较热门的视觉大模型，我们近期也会提供出 oneflow 的 vit 模型库以及相关的 benchmark 数据，非常欢迎大家到时候来体验 oneflow ~ 😄

**Comment by yuanms2** [comment]
> > nn.Graph 的 input 会稍稍影响流水，因为涉及到数据从 Python 传递到 Graph 内部，所以 Graph 的更高效做法是把 input 放在 Graph 内部，具体修改是：
> 
> 参考了这个修改之后，3090 上的速度为 395.83 img/s， 加速明显。但是我看了一下文档，按我的理解，只有基于 `flow.nn.OfrecordReader` 的数据读取方式可以作为静态图节点，基于 `flow.utils.data.DataLoader` 有办法优化吗？

做这个修改之后， nn graph 在3090 速度可以提升1倍？  再确认一下

**Comment by SunDoge** [comment]
> > > nn.Graph 的 input 会稍稍影响流水，因为涉及到数据从 Python 传递到 Graph 内部，所以 Graph 的更高效做法是把 input 放在 Graph 内部，具体修改是：
> > 
> > 
> > 参考了这个修改之后，3090 上的速度为 395.83 img/s， 加速明显。但是我看了一下文档，按我的理解，只有基于 `flow.nn.OfrecordReader` 的数据读取方式可以作为静态图节点，基于 `flow.utils.data.DataLoader` 有办法优化吗？
> 
> 做这个修改之后， nn graph 在3090 速度可以提升1倍？ 再确认一下

检查了下代码，batch size写错了，实际速度应为一半，即 197.915 img/s 。

**Comment by BBuf** [comment]
> > > > nn.Graph 的 input 会稍稍影响流水，因为涉及到数据从 Python 传递到 Graph 内部，所以 Graph 的更高效做法是把 input 放在 Graph 内部，具体修改是：
> > > 
> > > 
> > > 参考了这个修改之后，3090 上的速度为 395.83 img/s， 加速明显。但是我看了一下文档，按我的理解，只有基于 `flow.nn.OfrecordReader` 的数据读取方式可以作为静态图节点，基于 `flow.utils.data.DataLoader` 有办法优化吗？
> > 
> > 
> > 做这个修改之后， nn graph 在3090 速度可以提升1倍？ 再确认一下
> 
> 检查了下代码，batch size写错了，实际速度应为一半，即 197.915 img/s 。

方便的话可以更新一下oneflow vit模型的代码然后再测试一下速度，这里对Attention部分的写法做了一点优化。https://github.com/Oneflow-Inc/vision/blob/main/flowvision/models/vision_transformer.py

**Comment by SunDoge** [comment]
Hi, @BBuf
 
使用新版 vit 代码

```
100%|█████████████████████████████████████████| 320/320 [00:06<00:00, 45.95it/s]
100%|██████████████████████████████████████| 3200/3200 [00:16<00:00, 199.99it/s]
```

实验室 3090 非常抢手，很难找到机会做实验。

**Comment by BBuf** [comment]
> Hi, @BBuf
> 
> 使用新版 vit 代码
> 
> ```
> 100%|█████████████████████████████████████████| 320/320 [00:06<00:00, 45.95it/s]
> 100%|██████████████████████████████████████| 3200/3200 [00:16<00:00, 199.99it/s]
> ```
> 
> 实验室 3090 非常抢手，很难找到机会做实验。

好的，谢谢你。

**Comment by strint** [comment]
> Hi, @BBuf
> 
> 使用新版 vit 代码
> 
> ```
> 100%|█████████████████████████████████████████| 320/320 [00:06<00:00, 45.95it/s]
> 100%|██████████████████████████████████████| 3200/3200 [00:16<00:00, 199.99it/s]
> ```
> 
> 实验室 3090 非常抢手，很难找到机会做实验。

这结果是表示现在graph的结果和torch相近：199 vs 200 ？

**Comment by SunDoge** [comment]
我发现 `batch size` 对速度有一定影响，`batch size = 16, 32` 时，速度为 199 img/s, 但是 `batch size = 64, 128` 时速度会降到 175 img/s 。可能是 `oneflow` 同步机制的问题。

**Comment by BBuf** [comment]
你好，我们最终确认是因为cuda版本导致cublas调用时选择kernel不同的原因造成的，如果你在3090上编译一下cuda11.4版本的oneflow，会发现OneFlow仍比PyTorch快4%-5%左右。这是我的测试结果：

### 3090 单卡速度测试 （Master分支，源码编译cuda11.4版本）

- BatchSize = 64

| 框架模式      | Iter数 | 训练速度 |
| ------------- | ------ | -------- |
| PyTorch       | 200    | 63s      |
| OneFlow Eager | 200    | 61s      |
| OneFlow Graph | 200    | 60s      |

**Comment by SunDoge** [comment]
Thx.
